feat: #9 JWT 관련 컴포넌트/로그아웃 API 개발

wda067 opened the pull request:
### 📌 PR 타입(하나 이상의 PR 타입을 선택해주세요)
- 기능 추가

### 📌 관련 이슈
#9

### ✨ 반영 브랜치
feat/8 -> develop

### 📝 변경 사항
- JwtProvider / JwtService 등 JWT 컴포넌트 구현
- JWT 인증 필터 추가 및 401/403 예외 처리
- Access / Refresh Token 발급 및 재발급 로직 구현
- Refresh Token Redis 저장 및 검증 로직 추가
- 로그아웃 API 구현 (Refresh Token 삭제)

### ➕ 추가 메모

[`application-local.yml`](https://www.notion.so/2c4dc60fe5e28007955bc58412ef47d0?p=2bfdc60fe5e280f398c9cc522045c2dd&pm=s) 수정

### 🐛 테스트 결과

#### 소셜 로그인 요청 (기존 회원)
```
POST http://localhost:8080/api/auth/login
Content-Type: application/json

{
  "socialProvider": "KAKAO",
  "socialId": "existing-social-id-123",
  "email": "existing@example.com"
}
```

#### 응답 결과

```
HTTP/1.1 200 
Content-Type: application/json

{
  "code": "SUCCESS",
  "message": "요청이 성공적으로 처리되었습니다.",
  "data": {
    "signupRequired": false,
    "tokens": {
      "accessToken": "eyJhbGciOiJIUzUxMiJ9.eyJpc3MiOiJzd3lwIiwic3ViIjoiNCIsInR5cCI6ImFjY2VzcyIsImlhdCI6MTc2NTY0MTAwMiwiZXhwIjoxNzY1NjQxMDEyfQ.q2ld00-Am2G0DBw7wt7PYr_xR7U3p2O2vKVFVcgBpVd3qo8ROfvkaBFhmF7LymYRQU7q-Zxeb3h-zRRr_lVPoA",
      "refreshToken": "eyJhbGciOiJIUzUxMiJ9.eyJpc3MiOiJzd3lwIiwic3ViIjoiNCIsInR5cCI6InJlZnJlc2giLCJpYXQiOjE3NjU2NDEwMDIsImV4cCI6MTc2NjI0NTgwMn0.SIwx2YzJth-UZy8NzcXwbGrsOmt_PNTkpF6EgxB8ur8ADmX-uyA0a6tddzGFaLMwp0w8EJuJ3BmslwUwGf81AQ"
    }
  }
}
```

#### 회원가입 요청 (소셜 정보 + 유저 정보)

```
POST http://localhost:8080/api/auth/signup
Content-Type: application/json

{
  "login": {
    "socialProvider": "NAVER",
    "socialId": "new-social-id-999",
    "email": "newuser@example.com"
  },
  "nickname": "홍길동",
  "birthDate": "1998-01-03",
  "gender": "MALE"
}
```

#### 응답 결과

```
HTTP/1.1 200 
Content-Type: application/json

{
  "code": "SUCCESS",
  "message": "요청이 성공적으로 처리되었습니다.",
  "data": {
    "signupRequired": false,
    "tokens": {
      "accessToken": "eyJhbGciOiJIUzUxMiJ9.eyJpc3MiOiJzd3lwIiwic3ViIjoiNyIsInR5cCI6ImFjY2VzcyIsImlhdCI6MTc2NTY0MTA2NSwiZXhwIjoxNzY1NjQxMDc1fQ.mZkRP7V7ykXw-dHlrB6meeDQKTM_FodzJcZZjPK-scLmNFZAHe5U8naspAykQHXUeq3yhfeqlMBJR3vZrmJm6Q",
      "refreshToken": "eyJhbGciOiJIUzUxMiJ9.eyJpc3MiOiJzd3lwIiwic3ViIjoiNyIsInR5cCI6InJlZnJlc2giLCJpYXQiOjE3NjU2NDEwNjUsImV4cCI6MTc2NjI0NTg2NX0.AM3-Qtdqd58SPrZUd-4laKxy72z213YnY21mjPOKqUKfZ9GfjadQmGl2Ly61DH6H4wPgRPYExgkMrtwtGACQOg"
    }
  }
}
```

#### 로그아웃
```
POST http://localhost:8080/api/auth/logout
Authorization: Bearer {{accessToken}}
```

#### 응답 결과
```
HTTP/1.1 200 
Content-Type: application/json

{
  "code": "SUCCESS",
  "message": "요청이 성공적으로 처리되었습니다."
}
```

#### 토큰 재발급 요청 (Access Token 만료 시)

```
POST http://localhost:8080/api/auth/reissue
Content-Type: application/json

{
  "refreshToken": "{{refreshToken}}"
}
```

#### 응답 결과

```
HTTP/1.1 200 
Content-Type: application/json

{
  "code": "SUCCESS",
  "message": "요청이 성공적으로 처리되었습니다.",
  "data": {
    "accessToken": "eyJhbGciOiJIUzUxMiJ9.eyJpc3MiOiJzd3lwIiwic3ViIjoiNCIsInR5cCI6ImFjY2VzcyIsImlhdCI6MTc2NTY0MTIzMSwiZXhwIjoxNzY1NjQxMjQxfQ.6fmd76RcWiIx4aPHqzDrtHxreeP8yWg3QEsXUAub2-6P6TUNUfgYLpjEthLyew9IZBunvJ5rUTAZl-jPqzpjPg",
    "refreshToken": "eyJhbGciOiJIUzUxMiJ9.eyJpc3MiOiJzd3lwIiwic3ViIjoiNCIsInR5cCI6InJlZnJlc2giLCJpYXQiOjE3NjU2NDEyMzEsImV4cCI6MTc2NjI0NjAzMX0.mYQSxvV4A7ZQpT8c1rcl2vfSDJmyXSB6n4qsv21ym4533Wo27244md3RtXAX24XHP8JSFYjLquDi7E6eumbYRw"
  }
}
```

#### Access Token 만료 시 응답 결과
```
HTTP/1.1 401 
Content-Type: application/json;charset=UTF-8

{
  "code": "AUTH0001",
  "message": "인증이 필요합니다."
}
```

wda067 committed:
feat: #9 JWT 관련 컴포넌트/로그아웃 API 개발

- JwtProvider / JwtService 등 JWT 컴포넌트 구현
- JWT 인증 필터 추가 및 401/403 예외 처리
- Access / Refresh Token 발급 및 재발급 로직 구현
- Refresh Token Redis 저장 및 검증 로직 추가
- 로그아웃 API 구현 (Refresh Token 삭제)

roomie00 commented on src/main/java/pyws/swyp/auth/repository/RefreshTokenRepository.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.concurrent.TimeUnit;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.redis.core.StringRedisTemplate;
+import org.springframework.stereotype.Repository;
+import pyws.swyp.global.jwt.JwtProperties;
+import pyws.swyp.global.jwt.TokenHasher;
+
+@Repository
+@RequiredArgsConstructor
+public class RefreshTokenRepository {
+
+    private static final String PREFIX = "refresh:";
+
+    private final JwtProperties props;
+    private final StringRedisTemplate redisTemplate;
+
+    public void save(Long memberId, String refreshToken) {
+        String hashed = TokenHasher.hash(refreshToken);
+        redisTemplate.opsForValue().set(key(memberId), hashed, props.getRefreshTokenTtlMs(), TimeUnit.MILLISECONDS);
+    }
+
+    public String find(Long memberId) {
+        return redisTemplate.opsForValue().get(key(memberId));
+    }
+
+    public void delete(Long memberId) {
+        redisTemplate.delete(key(memberId));
+    }
+
+    public boolean matches(Long memberId, String refreshToken) {
+        String hashed = TokenHasher.hash(refreshToken);
+        String storedHashed = find(memberId);
+        return hashed.equals(storedHashed);
+    }
+
+    private String key(Long memberId) {
+        return PREFIX + memberId;
+    }
+}
+
```
redis 관련 코드의 재사용성을 위해 global 패키지 안에 component로 빼는 것은 어떨까요?

```
@Component
@RequiredArgsConstructor
public class RedisComponent {

    private final StringRedisTemplate redisTemplate;
    private final ObjectMapper objectMapper;

    public <DATA> void storeToRedis(String prefix, String key, long expirationMills, DATA data) {
        try {
            String json = objectMapper.writeValueAsString(data);
            redisTemplate.opsForValue().set(prefix + key, json, expirationMills, TimeUnit.MILLISECONDS);
        } catch (Exception e) {
            throw new CustomException(ErrorCode.REDIS_STORE_ERROR);
        }
    }

    public <DATA> DATA retrieveFromRedis(String prefix, String key, Class<DATA> clazz) {
        try {
            String savedCode = redisTemplate.opsForValue().get(prefix + key);

            if (savedCode != null && !savedCode.isBlank()) {
                return objectMapper.readValue(savedCode, clazz);
            } else return null;

        } catch (Exception e) {
            throw new CustomException(ErrorCode.REDIS_RETRIEVE_ERROR);
        }
    }
}
```
-> 이런식으로요!

wda067 replied:
그런 식으로 따로 빼도 괜찮을 거 같긴 한데
String을 ""로 다시 감싸서 저장을 해야 되나 싶긴 하네요..
아니면 String, JSON 구분해서 공통 메서드로 추가해두겠습니다!

wda067 committed:
feat: #9 RedisComponent 추가

- StringRedisTemplate 기반 Redis 공통 컴포넌트 구현
- String / JSON 데이터 저장·조회 메서드 추가

wda067 committed:
chore: #9 불필요한 애노테이션 제거

wda067 committed:
feat: #9 ObjectMapper 싱글톤 빈 등록

wda067 committed:
test: #9 RedisComponent 테스트

- RedisComponent 단위 테스트 추가
- String / JSON 저장·조회 로직 검증
- TTL 만료 동작 테스트
- Redis Testcontainers 구성

wda067 committed:
chore: #9 오타 수정

roomie00 commented on src/test/java/pyws/swyp/global/component/RedisComponentTest.java:
```diff
@@ -0,0 +1,108 @@
+import org.springframework.context.annotation.Import;
+import pyws.swyp.config.TestRedisConfig;
+
+@SpringBootTest
```
@SpringBootTest로 들어가게 되면 해당 파일에서 사용되지 않더라도 모든 bean을 다 로드하게 되고, 이로 인해 테스트 코드가 많아질수록 테스트 돌아가는게 느려질 수 있어서 아래와 같은 방식이 좋지 않을까 싶습니다 :)

```
// 옵션 1
@SpringBootTest(classes = {RedisComponent.class, TestRedisConfig.class}) 
// 옵션 2
@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = TestRedisConfig.class)
```

물론 앞으로 다른 코드들도 다 @SpringBootTest 통합테스트를 하실거라면 괜찮을 것 같습니다.
참고로 저는 Service 단은 @Mock 단위테스트, Controller 단은 @WebMvcTest로 슬라이스테스트(통합과 단위 사이...?) 이렇게 사용하고 있습니다.

저희가 시간 관계 상 테스트 코드에 대한 내용을 합의하지 않았었고, 각자 편한 방식대로 하면 된다고 생각해서 일단 approve 하겠습니다!! ☺️

wda067 replied:
전 보통 `@SpringBootTest`를 그냥 쓰긴 했었는데
```java
@SpringBootTest
@AutoConfigureMockMvc
class MemberControllerTest {
```
그러면 제안해주신 방법 적용해서 다시 올려놓겠습니다!

roomie00 replied:
앗 강요는 아니에요! 명근님이 @SpringBootTest가 편하시고 앞으로도 다 통합테스트로 진행하신다면 상관 없습니다 :)

wda067 replied:
아 넵..! 그럼 그대로 진행하겠습니다.
고치려니까 테스트가 계속 터지네요..

wda067 committed:
Merge branch 'develop' into feat/8
